src/v/kafka/server/handlers/fetch.cc

```diff
@@ -702,6 +711,186 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+      , _d_coeff_binding{config::shard_local_cfg().fetch_pid_d_coeff.bind()}
+      , _target_fetch_sg_util_binding{config::shard_local_cfg()
+                                        .fetch_pid_target_fetch_sg.bind()}
+      , _max_delay_binding{
+          config::shard_local_cfg().fetch_pid_max_debounce.bind()} {
+        _p_coeff_binding.watch([this] {
+            reset();
+            _p_coeff = config::shard_local_cfg().fetch_pid_p_coeff();
+        });
+        _i_coeff_binding.watch([this] {
+            reset();
+            _i_coeff = config::shard_local_cfg().fetch_pid_i_coeff();
+        });
+        _d_coeff_binding.watch([this] {
+            reset();
+            _d_coeff = config::shard_local_cfg().fetch_pid_d_coeff();
+        });
+        _target_fetch_sg_util_binding.watch([this] {
+            reset();
+            _target_fetch_sg_util
+              = scaling_factor
+                * config::shard_local_cfg().fetch_pid_target_fetch_sg();
+        });
+        _max_delay_binding.watch([this] {
+            _max_delay = config::shard_local_cfg().fetch_pid_max_debounce();
+        });
+        setup_metrics();
+    }
+
+    std::chrono::milliseconds current_delay(const ss::scheduling_group& sg) {
+        auto sampled_time = ss::sched_clock::now();
+        auto sample = sg.get_stats().runtime;
+        auto total_busy_time = ss::engine().total_busy_time();
+
+        auto dt = sampled_time - _last_sampled_time;
+        if (dt < sample_rate) {
+            return _last_delay;
+        }
+
+        auto busy_de = total_busy_time - _last_total_busy_time;
+        if (busy_de < 0 * duration_unit) {
+            return _last_delay;
+        }
+        int64_t busy = (scaling_factor * busy_de) / dt;
+        if (busy > scaling_factor) {
+            return _last_delay;
+        }
+
+        auto de = sample - _last_sample;
+        if (de < 0 * duration_unit) {
+            return _last_delay;
+        }
+        int64_t current_runtime = (scaling_factor * de) / dt;
+        if (current_runtime > scaling_factor) {
+            return _last_delay;
+        }
+
+        // If reactor utilization ~ 100% then ensure the fetch runtime is
+        // less than or equal to _target_fetch_sg_util.
+        //
+        // Otherwise ensure that reactor utilization is 100% even if the
+        // fetch runtime is above _target_fetch_sg_util.
+        constexpr int64_t target_reactor_util = 9990;
+        int64_t error = (busy < target_reactor_util)
+                          ? (busy - target_reactor_util)
+                          : (current_runtime - _target_fetch_sg_util);
+
+        // calculate the proportional
+        int64_t p = error / _p_coeff;
+
+        // calculate the integral
+        _error_int += error * (dt / duration_unit);
+        // Letting the integral fall below 0 only reduces the responsiveness
+        // in cases of sudden load spikes.
+        _error_int = std::max(0L, _error_int);
```

Comment:
Seems like integral is maybe just wrong for this controller? Since it's doing control on part of the interval, i.e., don't let CPU exceed a certain value, but don't drive _up_ CPU to the value if the natural load is below the target, plus we can't apply negative control values, etc. So it seems like the error can be non-zero for an arbitrary period of time so the error can get arbitrarily large (only in the positive direction, due to this cap), and then when the situation changes, the I term can overwhelm the other two for a long period of time.

Reply:
Though I guess the I term actually does a lot of the heavy lifting here since the P coefficient can't really be calculated exactly for different loads.

Reply:
The integral is rather important for the controller. Without the integral the controller tends to get "stuck" near, but not at, the target reactor utilization. I have added an upper bound to the integral as per your suggestion though.

Reply:
Thanks!